src/dataflow/src/source/s3.rs

```diff
@@ -663,9 +663,14 @@ enum S3Error {
         key: String,
         err: SdkError<GetObjectError>,
     },
-    ListObjectsFailed(SdkError<ListObjectsV2Error>),
-    RetryFailed,
-    IoError(std::io::Error),
+    ListObjectsFailed {
+        bucket: String,
+        err: SdkError<ListObjectsV2Error>,
+    },
+    IoError {
+        bucket: String,
+        err: std::io::Error,
+    },
 }
 
 impl std::error::Error for S3Error {}
```

Comment:
I think it would be nice to implement `source` on the error impl, for the cases that have inner errors (which right now is all of them), maybe a good TODO?